### DISCUSSION TITLE
x-bind broken/changed in AlpineJS > 3.7.1

### DISCUSSION THREAD

**Discussion by bep** [alpinejs/alpine#2942] [7. Bugs (archived)] [answered]
See https://codepen.io/bepsays/pen/zYRweYj

I have some AlpineJS sites that broke when upgrading to 3.8. They have x-bind constructs similar to the below:

```js
<!DOCTYPE html>
<html>
<script defer src="https://unpkg.com/alpinejs@3.7.1/dist/cdn.min.js"></script>
<body x-data="foo">
  <style>
    .red {
      color: red;
    }
    .blue {
      color: blue;
    }
  </style>
  <button type="button" x-bind="myClass()" x-on:click="clickMe()">Click Me</button>
  <script>
    document.addEventListener('alpine:init', () => {
      Alpine.data('foo', () => ({
        counter: 0,
        myClass: function() {
          return {
            'class': this.counter % 2 == 0 ? 'blue' : 'red'
          }
        },
        clickMe() {
          this.counter++;
        },
      }))
    })
  </script>
</body>
</html>
```

In `3.7.1` this works as expected. In > 3.7.1, the above just fails silently. I have tried to read the change log to understand how to adapt ...

**Comment by SimoTod** [comment, marked as answer]
Caleb refactored x-bind in that release to create 2 new global helpers.

From a quick check, the object-based x-bind doesn't use effect now so normal attributes are not reactive anymore. Bound directives still have the inner effect however, for the same reason, variables that are immediately resolved to a primitive datatype by the external x-bind won't be reactive as well.

At the moment, it looks like it seems like the only option is to make sure that the external x-bind returns a callback and assign it to an `x-bind:attribute` property so it gets tracked correctly. 

Something like
```js
['x-bind:class']() { return this.counter % 2 == 0 ? 'blue' : 'red' }
```

**Comment by bep** [comment]
@SimoTod thanks, that works ... However, I have a similar problem with dynamic transitions, which also worked fine in 3.7.1. I have tried multiple variants of the below:

https://codepen.io/bepsays/pen/PoQKgxX

```html
<div x-data="{ 
  counter: 0,
  transition: {
        ['x-transition:enter']() {
             if (this.counter > 5) {
              return ''
             }
             return 'transition ease-out duration-1000'
         },
         ['x-transition:enter-start']() {
             if (this.counter > 5) {
              return ''
             }
             return 'opacity-0 scale-50'
         },
          ['x-transition:enter-end']() {
             if (this.counter > 5) {
              return ''
             }
             return 'opacity-100 scale-100'
         },
   },
 }">
  <button @click="counter++">Count</button>
  <div class="" x-show="counter % 2 == 0" x-bind="transition" x-text="counter"></div>
</div>
```
